### DISCUSSION TITLE
[Closed] Separate toolkit without depending on MAUI for .NET 6 native apps

### DISCUSSION THREAD

**Discussion by dimonovdd** [CommunityToolkit/Maui#33] [General]
This discussion is related to [issue #11 (Add from Xamarin.Essentials: `Contacts`, `FilePicker`, `MediaPicker` and `WebAuthenticator`)](https://github.com/CommunityToolkit/Maui/issues/11)

I think: If we want to cut up Essentials, then we need to create a separate repository or nuget for these APIs without dependence on MAUI for using with .NET 6 native apps. 

Also to avoid code duplication we need to make some things from Maui.Essentials `public` ([`Permissions.EnsureDeclared`](https://github.com/xamarin/Essentials/blob/83e667d0d158ea50aeb26082714cd11d9549506a/Xamarin.Essentials/Permissions/Permissions.shared.cs#L19), [`IntermediateActivity`](https://github.com/xamarin/Essentials/blob/83e667d0d158ea50aeb26082714cd11d9549506a/Xamarin.Essentials/Platform/Platform.android.cs#L331) and etc.)

We also need to think about: Which library will contributors be able to add new features? Toolkit or directly to Essentials?

**Comment by dimonovdd** [comment]
https://github.com/xamarin/Essentials/issues/1757  is an example using MediaPicker from Xamarin.Essentials in native Xamarin applications without Xamarin.Forms

Developers also want to use Maui.Essentials and its features without Maui https://github.com/dotnet/maui/issues/1779

**Comment by TheCodeTraveler** [comment]
There are couple points to address here.

### Code Duplication

There will not be any code duplication between .NET MAUI, Maui Essentials and the Maui Community Toolkit. Both .NET MAUI and Maui Essentials have granted `InternalsVisibleTo` to `CommunityToolkit.Maui`: https://github.com/dotnet/maui/pull/1970

### Required Use of Maui

None of the new features being added in #11 require the use of .NET MAUI. I.e. You will still be able to use `FilePicker`, `Contacts`, `MediaPicker` and `WebAuthenticator` in your native .NET iOS (aka Xamarin.iOS) and native .NET Android (aka Xamarin.Android) apps.

### Creating Essentials.Toolkit

We do not currently have plans to create a new toolkit for these APIs.

**Reply by dimonovdd** [reply]
this ToolKit use property [`<UseMaui>true</UseMaui>`](https://github.com/CommunityToolkit/Maui/blob/main/src/CommunityToolkit.Maui/CommunityToolkit.Maui.csproj#L6). This can increase the build time and add complexity to the linker. As I wrote above there are those who want to use only `<UseMauiEssentials>true</UseMauiEssentials>`

**Reply by pictos** [reply]
> This can increase the build time 

We need to measure that and see if the build time will be an issue.

> and add complexity to the linker

Why? The XCT is linker safer by default and we want to keep this behavior here. The code that we will bring from Essentials are all `static` so that will be very easy to work with linker

**Reply by dimonovdd** [reply]
Why then divide Maui into parts, if using everything is not different from using only Maui.Essentials?
 `<UseMaui>`  `<UseMauiEssentials>` `<UseMauiAssets>`
 Don't these properties have different meanings?

**Comment by dimonovdd** [comment]
I am not alone, other developers also do not like unnecessary dependencies
https://github.com/dotnet/maui/issues/2337

**Comment by dimonovdd** [comment]
> These features will remain in Maui.Essentials

[link](https://github.com/CommunityToolkit/Maui/issues/11#issuecomment-1006753852)

**Comment by TheCodeTraveler** [comment]
Closing Discussion now that https://github.com/CommunityToolkit/Maui/issues/11 is closed

**Comment by CliffAgius** [comment]
Closed as answered.
